presto-hive/src/main/java/io/prestosql/plugin/hive/SyncPartitionMetadataProcedure.java

```diff
@@ -0,0 +1,239 @@
+import static io.prestosql.plugin.hive.HiveMetadata.PRESTO_QUERY_ID_NAME;
+import static io.prestosql.plugin.hive.HivePartitionManager.extractPartitionValues;
+import static io.prestosql.spi.StandardErrorCode.INVALID_PROCEDURE_ARGUMENT;
+import static io.prestosql.spi.block.MethodHandleUtil.methodHandle;
+import static io.prestosql.spi.type.StandardTypes.VARCHAR;
+import static java.lang.String.format;
+import static java.util.Locale.ENGLISH;
+import static java.util.Objects.requireNonNull;
+
+public class SyncPartitionMetadataProcedure
+        implements Provider<Procedure>
+{
+    public enum SyncMode
+    {
+        ADD, DROP, FULL
+    }
+
+    private static final MethodHandle SYNC_PARTITION_METADATA = methodHandle(
+            SyncPartitionMetadataProcedure.class,
+            "syncPartitionMetadata",
+            ConnectorSession.class,
+            String.class,
+            String.class,
+            String.class);
+
+    private final Supplier<TransactionalMetadata> hiveMetadataFactory;
+    private final HdfsEnvironment hdfsEnvironment;
+
+    @Inject
+    public SyncPartitionMetadataProcedure(
+            Supplier<TransactionalMetadata> hiveMetadataFactory,
+            HdfsEnvironment hdfsEnvironment)
+    {
+        this.hiveMetadataFactory = requireNonNull(hiveMetadataFactory, "hiveMetadataFactory is null");
+        this.hdfsEnvironment = requireNonNull(hdfsEnvironment, "hdfsEnvironment is null");
+    }
+
+    @Override
+    public Procedure get()
+    {
+        return new Procedure(
+                "system",
+                "sync_partition_metadata",
+                ImmutableList.of(
+                        new Argument("schema_name", VARCHAR),
+                        new Argument("table_name", VARCHAR),
+                        new Argument("mode", VARCHAR)),
+                SYNC_PARTITION_METADATA.bindTo(this));
+    }
+
+    public void syncPartitionMetadata(ConnectorSession session, String schemaName, String tableName, String mode)
+    {
+        SyncMode syncMode = toSyncMode(mode);
+        HdfsEnvironment.HdfsContext context = new HdfsEnvironment.HdfsContext(session, schemaName, tableName);
+        SemiTransactionalHiveMetastore metastore = ((HiveMetadata) hiveMetadataFactory.get()).getMetastore();
+        SchemaTableName schemaTableName = new SchemaTableName(schemaName, tableName);
+
+        Table table = metastore.getTable(schemaName, tableName).orElseThrow(() -> new TableNotFoundException(schemaTableName));
+        if (table.getPartitionColumns().isEmpty()) {
+            throw new PrestoException(INVALID_PROCEDURE_ARGUMENT, format("Table is not partitioned: %s.%s", schemaName, tableName));
+        }
+        Path defaultLocation = new Path(table.getStorage().getLocation());
+
+        Set<String> partitionsToAdd;
+        Set<String> partitionsToDrop;
+
+        try (FileSystem fileSystem = hdfsEnvironment.getFileSystem(context, defaultLocation)) {
```

Comment:
This is wrong and is the cause of the strange errors you saw. Hadoop `FileSystem` is `Closeable`, so you would expect that you could close it. But they are actually global singletons, so if you close it, you mess things up permanently for everyone. (it's a great design...)

Move this to be a normal assignment the try block. After that, you should be able to move `syncPartitions` back outside the block.